numpy/core/src/umath/ufunc_object.c

```diff
@@ -3061,32 +3063,44 @@ PyUFunc_Reduce(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
     if (identity == NULL) {
         return NULL;
     }
+
+    /* Get the initial value */
+    if (initial == NULL) {
+        initial = identity;
+        initial_given = 0;
+    }
+    else {
+        Py_DECREF(identity);
+        Py_INCREF(initial);  /* match the reference count in the if above */
+        initial_given = 1;
+    }
+
     /*
      * The identity for a dynamic dtype like
      * object arrays can't be used in general
      */
-    if (identity != Py_None && PyArray_ISOBJECT(arr) && PyArray_SIZE(arr) != 0) {
-        Py_DECREF(identity);
-        identity = Py_None;
-        Py_INCREF(identity);
+    if (!initial_given && initial != Py_None && PyArray_ISOBJECT(arr) && PyArray_SIZE(arr) != 0) {
```

Comment:
I would move this if statement inside the `if (initial == NULL)` clause above (and then test `identity`), so that you can do without the `initial_given` variable.